mirrord/layer/src/socket/ops.rs

```diff
@@ -765,13 +774,96 @@ pub(super) fn gethostname() -> Detour<&'static CString> {
     HOSTNAME.get_or_detour_init(remote_hostname_string)
 }
 
-/// libraries like `c-ares` expect messages to be from the address they were sent to.
-/// currently this function just fills in the address expected by the caller.
-#[tracing::instrument(level = "trace")]
-pub(super) fn recv_from(sockfd: i32, sockaddr: *mut sockaddr, addrlen: *mut u32) -> Detour<()> {
-    let socket_state = SOCKETS.get(&sockfd)?.clone();
-    if let SocketState::Connected(Connected { remote_address, .. }) = &socket_state.state {
-        fill_address(sockaddr, addrlen, (remote_address.clone()).try_into()?)?;
-    }
-    Detour::Success(())
+/// We handle UDP sockets by putting them in a sort of _semantically_ connected state, meaning that
+/// we don't actually [`libc::connect`] `sockfd`, but we change the respective [`UserSocket`] to
+/// [`SocketState::Connected`].
+///
+/// Here we check for this invariant, so if a user calls [`libc::recv_from`] without previously
+/// either connecting this `sockfd`, or calling [`libc::send_to`], we're going to panic.
+///
+/// It performs the [`fill_address`] requirement of [`libc::recv_from`] with the correct remote
+/// address (instead of using our interceptor address).
+#[tracing::instrument(level = "trace", ret, skip(out_buffer, raw_source, source_length))]
+pub(super) fn recv_from(
+    sockfd: i32,
+    out_buffer: *mut c_void,
+    buffer_length: size_t,
+    flags: i32,
+    raw_source: *mut sockaddr,
+    source_length: *mut socklen_t,
+) -> Detour<isize> {
+    let address = SOCKETS
+        .get(&sockfd)
+        .map(|socket| match &socket.state {
+            SocketState::Connected(connected) => connected.remote_address.clone(),
+            _ => unreachable!(),
+        })
+        .map(SocketAddress::try_into)??;
+
+    let recv_from_result = unsafe {
+        FN_RECV_FROM(
+            sockfd,
+            out_buffer,
+            buffer_length,
+            flags,
+            raw_source,
+            source_length,
+        )
+    };
+
+    fill_address(raw_source, source_length, address)?;
+
+    Detour::Success(recv_from_result)
+}
+
+/// There is a bit of trickery going on here, as this function first triggers a _semantical_
+/// connection to an interceptor socket (we don't [`libc::connect`] this `sockfd`, just change the
+/// [`UserSocket`] state), and only then calls the actual [`libc::send_to`] to send `raw_message` to
+/// the interceptor address (instead of the `raw_destination`).
+///
+/// See [`recv_from`] for more information.
+#[tracing::instrument(
+    level = "trace",
+    ret,
+    skip(raw_message, raw_destination, destination_length)
+)]
+pub(super) fn send_to(
+    sockfd: RawFd,
+    raw_message: *const c_void,
+    message_length: usize,
+    flags: i32,
+    raw_destination: *const sockaddr,
+    destination_length: socklen_t,
+) -> Detour<isize> {
+    let destination = SockAddr::try_from_raw(raw_destination, destination_length)?;
```

Comment:
we should add a test when we change addresses inbetween send_to with recv_from from both to see that it's working as well

Reply:
if it doesn't work maybe we should make this implementation available only for dns (port 53) to avoid introducing bugs in the apps

Reply:
Alright, now that the basic integration test is working, I'll try testing this out.

Should this patch be part of `MIRRORD_UDP_OUTGOING` or `MIRRORD_DNS` config? I can see an user wanting dns resolution, but disabling udp, so I'm not sure where to gate this.

Reply:
Quick test shows that we end up blocking on the second `sendto`:

```sh
sent 10 to 127.0.0.1:8888
2023-06-13T21:01:18.363645Z TRACE ThreadId(12) interceptor_task: mirrord_layer::outgoing::udp: interceptor_task -> Received data from user socket 127.0.0.1:41947 connection_id=0 mirror_socket=PollEvented { io: Some(UdpSocket { addr: 127.0.0.1:52083, fd: 13 }) }

(logs showing that stuff is working ...)

sent 10 to 127.0.0.1:9999
2023-06-13T21:01:18.408825Z TRACE ThreadId(19) interceptor_task: mirrord_layer::outgoing::udp: interceptor_task -> Received data from user socket 127.0.0.1:41947 connection_id=1 mirror_socket=PollEvented { io: Some(UdpSocket { addr: 127.0.0.1:32931, fd: 14 }) }

2023-06-13T21:01:18.408869Z TRACE ThreadId(01) recv_from: mirrord_layer::socket::ops: new sockfd=10 buffer_length=20 flags=0

(we never receive anything, `recv_from` just blocks)

2023-06-13T21:06:48.472229Z TRACE ThreadId(01) recv_from: mirrord_layer::socket::ops: new sockfd=10 buffer_length=20 flags=0
2023-06-13T21:06:48.472275Z TRACE ThreadId(19) next: mirrord_layer::tcp_steal: new
2023-06-13T21:06:48.472287Z TRACE ThreadId(19) next: mirrord_layer::tcp_steal: close time.busy=1.98µs time.idle=10.4µs
2023-06-13T21:06:48.472306Z TRACE ThreadId(19) mirrord_layer::connection: Sending client message: UdpOutgoing(Write(LayerWrite { connection_id: 1, bytes (length): 10 }))

(we stay here forever)
```

Reply:
Wow, what a wild ride this became:

1. `fill_address`:

To properly `fill_address` of a `recvfrom` address, we have to check if this address belongs to us, due to how we change a bound address in `bind`.

2. port 0:

Now that we have port 0 enabled to `bind` for UDP, when the user calls `getsockname` they should get a valid address + port combo, `bind(127.0.0.1:0)` has to become address `127.0.0.1:{not 0}` (this is what happens without mirrord).

3. `sendto` and `recvfrom`:

These functions now have to check if the address they're sending/receiving is an actual address, that we can send packets on.